### ISSUE TITLE
[$2000] Add option to reset Expensify password from new bank account setup pane

### ISSUE THREAD

**Issue by stephanieelliott** [Expensify/App#6877]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Log into new.expensify.com and navigate to profile icon > Payments > Add Payment Method
2. Click `Bank Account`
3. Log into any bank account via Plaid
4. Select bank account from dropdown menu to see `Enter Expensify password` field appear. This is where we want to add the new option to reset password. 

## Desired Result:

On the page where `Enter Expensify password` appears, add an option to reset the Expensify password. It should appear as a blue hyperlink which says `Forgot? Get a magic sign-in link!`. 

## Actual Result:
Currently there is no option to reset password on that page (described in Step 4). 

## Workaround:
unknown
## Platform:

Where is this issue occurring?

- All

**Version Number:** v1.1.22-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Notes/Photos/Videos:** Any additional supporting documentation
![image](https://user-images.githubusercontent.com/31225194/147119908-6b346b5e-b9db-4957-947a-3e26e44e42ad.png)

**Expensify/Expensify Issue URL:** https://github.com/Expensify/Expensify/issues/188564
**Issue reported by:** Internal
**Slack conversation:**  https://expensify.slack.com/archives/C01966DBVLL/p1639082498315100?thread_ts=1638979187.291600&cid=C01966DBVLL

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by stephanieelliott** [comment]
Posted to Upwork: https://www.upwork.com/jobs/~01087482f69cffd1bb

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @parasharrajat (`Exported`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @chiragsalian (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by sethraj14** [comment]
@stephanieelliott Proposal 
I would add this code bit for adding the link of resetting and magic link- 
`
                    <View style={[styles.changeExpensifyLoginLinkContainer]}>
                        <TouchableOpacity
                            style={[styles.mt2]}
                            onPress={Session.resetPassword}
                            underlayColor={themeColors.componentBG}
                        >
                            <ExpensifyText style={[styles.link]}>
                                {this.props.translate('passwordForm.forgot')}
                            </ExpensifyText>
                        </TouchableOpacity>
                    </View>
`

The component where this code needs to be added - `src/components/AddPlaidBankAccount.js`

**Comment by sethraj14** [comment]
In order to show you the working demo of the same, I would be needing test bank accounts if there is any/ or if there is some other way to get to that state.

**Comment by parasharrajat** [comment]
@stephanieelliott What should happen when the user clicks that link?

**Comment by mananjadhav** [comment]
**Proposal**

I've got a question on the feedback loop/navigation cycle after the link. But some pieces I've added in the proposal.

1. I will add a link that will be clickable
```
<TouchableOpacity
    style={[styles.mt2]}
    onPress={this.resetPassword}
    underlayColor={themeColors.componentBG}
>
    <ExpensifyText style={[styles.link]}>
        {this.props.translate('passwordForm.forgot')}
    </ExpensifyText>
</TouchableOpacity>

```

2. Add a function `resetPassword` which will internally call `Session.resetPassword`
```
    function resetPassword() {
       Session.resetPassword();

       // We will need to add feedback as well
    }

```

3. There needs to be a feedback too. In the Login section it redirects to the magic link page. I don't think that's the expected behavior here. We have a few options of showing the feedback:
    a. Show a growl message in the `AddPlaidBankAccount` component.
    b. Add an `ExpensifyText` component with the feedback message
    c. Add the growl message in the `Session.resetPassword` function


What I am not sure about is once the user clicks, what happens to the current page. If go to the email and then click on the magic link to reset the password, will the user be redirected to the `AddPlaidBankAccount` page?

**Comment by parasharrajat** [comment]
Sorry for the delay here in reviewing proposals. I am waiting for an answer to https://github.com/Expensify/App/issues/6877#issuecomment-999730666.

**Comment by stephanieelliott** [comment]
This is still being discussed in the Slack [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1640792952211900). Based on where the discussion is now, this job is likely to change so let's hold off on any more proposals for now.

**Comment by chiragsalian** [comment]
An update, the conversation is still active on slack.

**Comment by stephanieelliott** [comment]
Discussion still ongoing. A few different proposals at play, summarized [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1642718870250600?thread_ts=1640792952.211900&cid=C01GTK53T8Q).

**Comment by mallenexpensify** [comment]
Sorry for the delay y'all. Since the main use case will be people on web mobile and web desktop, I think we should focus on those (at least to start).  I also think we should focus on using our existing tools and flows vs creating something new. 

For both mobile and desktop web:
- User forgets password while on the `Bank Account` screen
- User clicks Magic link to get new password
- Growl is surfaced to user letting them know the email/text has been send and that the user needs to click the link to reset their pass
- User clicks magic link and is directed to the `setpassword` page. 
- User creates new password, clicks "Set password" then is redirected to where they previously left off at the bank account dropdown menu.
- User either needs to enter pass again or not since they just created a new one the step before (unsure of this) 

If the user has another tab opened on mobile or desktop web, they'll either be auto-signed in once the new password is created, they'll need to refresh to be signed in or they'll need to enter their new pass.  For mobile web, my guess is they won't ever specifically go back to the earlier open tab because they'll be able to complete the action they need to in their current tab.

**Comment by parasharrajat** [comment]
Thanks, @mallenexpensify Open to proposals...based on the above requirements.

**Comment by mallenexpensify** [comment]
Doubled price to $500 and created new job, old one had expired https://www.upwork.com/jobs/~0152b68b2bb7a2e6df

**Comment by mananjadhav** [comment]
My proposal here covers the requirement https://github.com/Expensify/App/issues/6877#issuecomment-999732939

@parasharrajat Can help me confirm if I am missing anything.

Need clarifications on:
>  then is redirected to where they previously left off at the bank account dropdown menu. (@AndrewGable confirmed this is possible - internal link)

Is this available by default or the current proposal needs to add it?

>  If the user has another tab opened on desktop web they'll be auto-signed in once the new password is created

We're saying the `setpassword` screen opens fine even if I am already logged in? Because if I do a password reset now? and I go to an existing tab, it redirects me to the homepage.

Or we're saying we need to update `SetPassword` page to allow opening even if I am already signedin?

**Comment by parasharrajat** [comment]
> Is this available by default or the current proposal needs to add it?

I can only think one way. use a redirect param in the magic link which we can use to redirect the user to the same page where the user clicked forgot pass.

> Or we're saying we need to update SetPassword page to allow opening even if I am already signedin?

I think yes otherwise redirection will not work. Magin link will not open if a user is a login. You will have to copy that screen from Public routes to AuthScreens and then open that page over RHN. Maybe add a new route added for this if needed. Apart from it Create a new page that wraps PasswordForm without sign-in page layout.

**Comment by mananjadhav** [comment]
> use a redirect param in the magic link

I agree, that's the standard clean way. Will have to test on native once. But my question was is it available at the moment? Didn't understand when we say Possible means exists or doable.

> Magic link will not open if a user is a login

Thanks for clarifying. I'll update the proposal to add these details.

**Comment by parasharrajat** [comment]
Didn't understand when we say Possible means exist or doable.

No idea, better to ping on the same slack thread.

**Comment by mallenexpensify** [comment]
> Is this available by default or the current proposal needs to add it?

@AndrewGable , can you help shine some light on "redirected to where they previously left off at the bank account dropdown menu" part [from here](https://github.com/Expensify/App/issues/6877#issuecomment-1026371826)?

> If the user has another tab opened on desktop web they'll be auto-signed in once the new password is created

For the proposal for this issue, this isn't part of the deliverable or consideration.  I was just pointing out there wasn't a downside with the existing tab because it'll remain signed in. 
 
Hope that helps ¯\_(ツ)_/¯

**Comment by AndrewGable** [comment]
I don't think redirecting to the bank account dropdown menu is possible, but I do not know if it's currently implemented or not.

**Comment by thesahindia** [comment]
> Hey @thesahindia are you still working updating your proposal, or is it ready for @parasharrajat to review?
I think Manan is working on this 
cc: @mananjadhav

**Comment by stephanieelliott** [comment]
Oops, sorry @thesahindia! 

Hey @mananjadhav are you still working updating your proposal, or have you updated it and it is ready for review?

**Comment by stephanieelliott** [comment]
Sounds like @mananjadhav may not be working on this proposal anymore. Raised price $1000
https://www.upwork.com/jobs/~0152b68b2bb7a2e6df

**Comment by MelvinBot** [comment]
Current assignee @stephanieelliott is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @puneetlath (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by stephanieelliott** [comment]
I'm about to go OOO, so reapplied the label to assign this to another CM team member (thanks @puneetlath!)

**Comment by puneetlath** [comment]
@joekaufmanexpensify @kevinksullivan what should the UX be here after someone clicks that "reset password" link?

**Comment by joekaufmanexpensify** [comment]
@mallenexpensify outlined the next steps for this change well [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1643678199019029?thread_ts=1640792952.211900&cid=C01GTK53T8Q). So once the user clicks the password link, we'll want to:

1. Show a growl indicating that the reset link has been sent. 
2. User clicks password reset link and is directed to change their password on `setPassword` page. 
3. User sets new password and is directed back to where they were in the bank account set up flow (if possible). 
4. User enters new password to complete bank account set up flow.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @mallenexpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by puneetlath** [comment]
>User sets new password and is directed back to where they were in the bank account set up flow (if possible).

@chiragsalian I think this part will need to be internal right? 

Anyways, @mallenexpensify I'm about to head out on my two-month sabbatical, so handing this off to you in case it does remain External. Thanks!

**Comment by chiragsalian** [comment]
Ah i guess we did get a bit sidetracked in this issue. I agree with @mallenexpensify that if we allowed validateCode to be entered manually then we should propose it as a new feature so for now let's not consider it. I like the steps proposed but how about we keep it simpler and not include step 3, only because atm we don't do more complex redirects in newDot (like redirecting to specific page once setPassword is complete). Its definitely possible but i'm not sure if its worth the gain here. 
So how about we just do,
1. Show password forgot message and link.
2. If user clicks it we trigger API command `ResetPassword ` just the same way as the login page and show the user a message to check their email for the magic link.
3. Once user clicks on the link they go to password rest page on the `SetPassword` page.
4. Once the new password is entered they are just logged into the app and can go to the bank account flow to resume.

All of the above can be handled externally.

If we would like the link to allow them to reset password and have it go to a specific page then i think its best to create a follow-up issue since its mostly internal and at this time it might be best on slack if a link with complex redirecting is better or if we should allow the user to manually enter a validateCode. I think i vote the latter because even with a more complex link it opens a new tab with a new page (SetPassword), while i think its a better user experience to continue from the page they are in (like bankAccount login, enter validate code and continue).

If others agree i think we should update the desired result section of the OP so that contributors can share their proposal. Also not sure why this is daily, bumping it down to weekly (Feel free to update if anyone disagrees).

**Comment by parasharrajat** [comment]
Sounds approachable to me @chiragsalian.

**Comment by mallenexpensify** [comment]
Thanks @chiragsalian !  I just doubled the price to $2000.  With [your steps listed above](https://github.com/Expensify/App/issues/6877#issuecomment-1061000993), hopefully we'll be able to get this fixed and put it to bed soon.

**Comment by mananjadhav** [comment]
I had posted a proposal here https://github.com/Expensify/App/issues/6877#issuecomment-999732939 which covered almost everything that is expected. I had one question on if there's any feedback/growl to be shown when the user clicks.

**Comment by thesahindia** [comment]
### Proposal 
We can add the code written below inside [AddPersonalBankAccountPage ](https://github.com/Expensify/App/blob/main/src%2Fpages%2FAddPersonalBankAccountPage.js)to show the option to reset the link 
```js 
<View style={[styles.changeExpensifyLoginLinkContainer]}>
                    <TouchableOpacity
                        style={[styles.mt2]}
                        onPress={resetPassword}
                    >
                        <Text style={[styles.link]}>
                            {props.translate('passwordForm.forgot')}
                        </Text>
                    </TouchableOpacity>
                </View>
```
We can use a state `isValidationLinkFormVisible` here when the user taps on `forgot` option we will call `Session.resetPassword` and will make the `isValidationLinkFormVisible` true and when `isValidationLinkFormVisible` is true will show the [ResendValidationForm ](https://github.com/Expensify/App/blob/main/src%2Fpages%2Fsignin%2FResendValidationForm.js) component, we currently use this for forgot password option. This way we won't need to show a growl as we trying to not use the growl as much as we can.

**Comment by parasharrajat** [comment]
Based on the new [requirements](https://github.com/Expensify/App/issues/6877#issuecomment-1061000993), @mananjadhav 's proposal looks better. 

But I have one question from all the contributors. I don't think that the app takes the user to the Setpassword page when the user is logined into the app. 

So after step 3, how will you handle that?

**Comment by mananjadhav** [comment]
> But I have one question from all the contributors. I don't think that the app takes the user to the Setpassword page when the user is logined into the app.

We discussed somewhere in the comments but here are two options:
1. We copy the same route in AuthScreens as well (not recommended as the layout will be SigninPageLayout)
2. We create a new route with a new component, that is docked on the right modal and uses `NewPasswordForm` component. We could even move the code to common component for `SetPasswordPage` and `AuthSetPasswordPage`

**Comment by parasharrajat** [comment]
@chiragsalian What are you thoughts on this :arrow_up: ?

**Comment by chiragsalian** [comment]
> I don't think that the app takes the user to the Setpassword page when the user is logined into the app.

Uff, good point. I like manana's suggested option 2, but yeah like manan said, at this point, we've to create a new route and component. Also, a bit of backend code to send a different route when the user is logged in. Hmm at this point I guess I'm wondering why not create a new component to accept validate code and no new route so the user will get the email to enter validateCode and then just continue where they left off. I guess both feel like new features to me and I'm not sure which one is the best.

I do like the route idea more but I feel like the pain point is I believe the link cannot directly open in the desktop app and if a user clicks on it will instead open a browser tab to continue their session there even if there is a desktop app open correct? I feel that's not an ideal user experience. Anyway curious to know if you guys have any thoughts on this?

Also @mallenexpensify, I was just curious about the ROI of this task. I'm mostly wondering if we feel like it's really worth it at this time because once logged in in oldDot we never really show forgot password anywhere unless I'm mistaken. Can you let me know if I'm mistaken or regardless of oldDot if we feel like this is a valuable flow for newDot? If it's valuable, it looks like the feature to enter your password is new and there are different flows (we would need a list of all the places this is there) where we have it so we might want to consider a more permanent solution.
![image](https://user-images.githubusercontent.com/14356145/157347409-519f19a3-8cda-499a-b380-06b89c0854e5.png)

**Comment by parasharrajat** [comment]
> Hmm at this point I guess I'm wondering why not create a new component to accept validate code and no new route so the user will get the email to enter validateCode and then just continue where they left off. 

I feel like this is the best thing to do. In both approaches, we need backend changes so I think this one is better to do. If we are planning to add forgot password link in multiple places then it makes sense.

**Comment by chiragsalian** [comment]
Took the discussion to [slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1647303519841519).

**Comment by mallenexpensify** [comment]
Not doubling, lot of discussion in link above

**Comment by mallenexpensify** [comment]
@chiragsalian is keeping the convo alive! Waiting til we have a clear path before moving forward

**Comment by mallenexpensify** [comment]
@chiragsalian , if/when you're back from OOO, can you provide and update plz?

**Comment by chiragsalian** [comment]
Yes, sry for the delay, the slack discussion resulted in a couple of thought but unfortunately no consensus. Discussed 1:1 with Matt about next steps yesterday. I feel its best to post in whats-next so that we can have more eyes on this to reach a decision. 

Before i post though i've asked bondy if he's free in the next few days to discuss the vulnerability report he received to determine how valuable the issue is at this time. Removing label external until we can figure out next steps.

**Comment by chiragsalian** [comment]
Discussed with @bondydaa, he mentions the original issue is not valid anymore and said that we moved authToken cookies to secure, http-only cookies so in theory an attacker shouldn't be able to get an authToken via an XSS exploit anymore. 

Closing issue for now. Bondy mentioned he'll talk to cole to figure out the next PEN testing and if the results show this feature is still needed then we'll decide the next steps here.

**Comment by mallenexpensify** [comment]
Checking internally about potential compensation for this issue since work/research/reviews have been done
https://expensify.slack.com/archives/C02NK2DQWUX/p1649707752228549

**Comment by mallenexpensify** [comment]
@parasharrajat does $500 seem fair for your C+ contributions on this issue the past few months?

**Comment by parasharrajat** [comment]
Yeah, that would be fine. Thanks.

**Comment by mallenexpensify** [comment]
Can you accept the offer here Rajat https://www.upwork.com/jobs/~0152b68b2bb7a2e6df
and post a comment when you have?

**Comment by parasharrajat** [comment]
I just accepted it.

**Comment by mallenexpensify** [comment]
Paid Rajat!
